Optimize remove unused columns in plan

antaljanosbenjamin opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Add optimization to remove unused columns in query plans. Resolves #75152

### Details
Resolves #75152

antaljanosbenjamin committed:
Fix typos

antaljanosbenjamin committed:
Make `ExpressionStep`, `FilterStep` and `JoinStepLogical` able to remove unused columns

antaljanosbenjamin committed:
Add tryRemoveUnusedColumns optimization

antaljanosbenjamin committed:
Add test

KochetovNicolai commented on src/Core/Settings.cpp:
```diff
@@ -5923,6 +5923,19 @@ This only affects operations performed on the client side, in particular parsing
 
 Note that initially (24.12) there was a server setting (`send_settings_to_client`), but latter it got replaced with this client setting, for better usability.
+)", 0) \
+    DECLARE(Bool, query_plan_try_remove_unused_columns, true, R"(
```
query_plan_remove_unused_columns

antaljanosbenjamin committed:
Fix build

antaljanosbenjamin commented:
Just to make it into public knowledge: the current approach is not completely okay, because it might break the rule of subsequent steps must have matching input-output headers. So the plan is to collect the extra columns in the plan from top to bottom, and then remove the ones that can be removed from bottom to the top.

antaljanosbenjamin committed:
Rename setting

antaljanosbenjamin committed:
Trying to make optimization work without breaking any queries

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix build

antaljanosbenjamin committed:
Fix settings changes history

antaljanosbenjamin committed:
Add debugExplainPlan for future debugging

antaljanosbenjamin committed:
Consider all output of post join actions

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Do not add filter column multiple times

antaljanosbenjamin committed:
Keep at least one output from pre join actions

antaljanosbenjamin committed:
Try to fix segfault

antaljanosbenjamin committed:
Handle removal of columns that are not handled by actions

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fixes after merge

antaljanosbenjamin committed:
Consider filter conditions too in joins

antaljanosbenjamin committed:
Avoid modifying actions when the required outputs are the same

antaljanosbenjamin committed:
Merge branch 'master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Don't remove unused column in case of duplicated inputs or outputs

ActionsDAG doesn't work well with duplicated input/output names, so let's ignore that for now.

antaljanosbenjamin committed:
Style fix

antaljanosbenjamin committed:
Choose better required output column in `JoinStepLogical` when there are no required columns

KochetovNicolai commented on src/Processors/QueryPlan/IQueryPlanStep.h:
```diff
@@ -113,6 +113,28 @@ class IQueryPlanStep
+    /// It is guaranteed that the output header of the step will contain all columns from
+    /// required_outputs and might contain some other columns too.
+    /// Can be used only if canRemoveUnusedColumns returns true.
+    /// If none of the output columns are removed, then it is preferable to keep the order
```
I think we can make it even stricter here. The order of the not removed columns should stay the same.

devcrafter replied:
I guess the original comment refers to the guarantee keeping the column order in case none is removed, or?

KochetovNicolai commented on src/Processors/QueryPlan/IQueryPlanStep.h:
```diff
@@ -113,6 +113,28 @@ class IQueryPlanStep
+    /// Can be used only if canRemoveUnusedColumns returns true.
+    /// If none of the output columns are removed, then it is preferable to keep the order
+    /// of output columns in order to avoid excessive updates of input headers in parent steps.
+    virtual UnusedColumnRemovalResult removeUnusedColumns(NameMultiSet /*required_outputs*/, bool /*remove_inputs*/);
```
Possible later, but we can make this function position-based.
Get a list of output positions that are not needed, and return two lists of in/out positions that were removed.

KochetovNicolai commented on src/Processors/QueryPlan/ExpressionStep.cpp:
```diff
@@ -97,6 +99,94 @@ std::unique_ptr<IQueryPlanStep> ExpressionStep::deserialize(Deserialization & ct
+
+IQueryPlanStep::UnusedColumnRemovalResult ExpressionStep::removeUnusedColumns(NameMultiSet required_outputs, bool remove_inputs)
+{
+    if (output_header == nullptr)
```
FilterStep -> ExpressionStep

antaljanosbenjamin committed:
Update tests to expect the optimized plan

antaljanosbenjamin committed:
Fix settings changes history

antaljanosbenjamin committed:
Fix pushDownFilter to not change output headers

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix child plan generation with aliases in StorageMerge

antaljanosbenjamin committed:
Do not add column to actions DAG when adding inputs back from input header

antaljanosbenjamin committed:
Add extra check to ExpressionStep

KochetovNicolai committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into optimize-remove-unused-columns-in-plan

KochetovNicolai committed:
Allow FilterDescription by numeric column.

KochetovNicolai committed:
Relax check in FilterTransform. Update some tests.

KochetovNicolai committed:
Update more tests.

antaljanosbenjamin committed:
Update even more tests

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Move new setting to 25.9

antaljanosbenjamin committed:
Keep at least one output from left and right actions in JoinStepLogical

antaljanosbenjamin committed:
Merge branch 'master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Remove unreachable inputs after merging DAGs

antaljanosbenjamin committed:
Restore output columns in pre join actions after merging filter into join condition

taiyang-li commented:
@antaljanosbenjamin I found there are some unused columns in MergeTreeSource. They are introduced by prewhere optimization. 

Consider a query like 
``` sql 
explain plan header = 1, actions = 1  
select account_id as accountId, country as country from ad_dsp_analyse_day_agg_v2 where day >= '2025-07-01' and day <= '2025-07-31' and source_type in ('1') and toUInt64OrZero(account_id) < 900000 and ad_id!='global'  FORMAT TSV  settings optimize_move_to_prewhere = true; 

Expression ((Project names + Projection))
Header: accountId String
        country LowCardinality(String)
Actions: INPUT : 0 -> __table1.account_id String : 0
         INPUT : 1 -> __table1.country LowCardinality(String) : 1
         ALIAS __table1.account_id :: 0 -> accountId String : 2
         ALIAS __table1.country :: 1 -> country LowCardinality(String) : 0
Positions: 2 0
  Expression ((WHERE + Change column names to column identifiers))
  Header: __table1.account_id String
          __table1.country LowCardinality(String)
          day Date
          source_type String
          ad_id String
  Actions: INPUT : 1 -> country LowCardinality(String) : 0
           INPUT : 0 -> account_id String : 1
           ALIAS country :: 0 -> __table1.country LowCardinality(String) : 2
           ALIAS account_id :: 1 -> __table1.account_id String : 0
  Positions: 0 2
    ReadFromMergeTree (default.ad_dsp_analyse_day_agg_v2)
    Header: account_id String
            day Date
            source_type String
            ad_id String
            country LowCardinality(String)
    ReadType: Default
    Parts: 4
    Granules: 104728
    Prewhere info
    Need filter: 1
      Prewhere filter
      Prewhere filter column: and(greaterOrEquals(__table1.day, \'2025-07-01\'_String), lessOrEquals(__table1.day, \'2025-07-31\'_String), in(__table1.source_type, __set_String_8953384307879712516_13388813323589638527), notEquals(__table1.ad_id, \'global\'_String), less(toUInt64OrZero(__table1.account_id), 900000_UInt32)) (removed)
      Actions: INPUT : 0 -> account_id String : 0
               INPUT : 1 -> day Date : 1
               INPUT : 2 -> source_type String : 2
               INPUT : 3 -> ad_id String : 3
               COLUMN Const(String) -> \'2025-07-01\'_String String : 4
               COLUMN Const(String) -> \'2025-07-31\'_String String : 5
               COLUMN Const(Set) -> __set_String_8953384307879712516_13388813323589638527 Set : 6
               COLUMN Const(UInt32) -> 900000_UInt32 UInt32 : 7
               COLUMN Const(String) -> \'global\'_String String : 8
               FUNCTION toUInt64OrZero(account_id : 0) -> toUInt64OrZero(__table1.account_id) UInt64 : 9
               FUNCTION greaterOrEquals(day : 1, \'2025-07-01\'_String :: 4) -> greaterOrEquals(__table1.day, \'2025-07-01\'_String) UInt8 : 10
               FUNCTION lessOrEquals(day : 1, \'2025-07-31\'_String :: 5) -> lessOrEquals(__table1.day, \'2025-07-31\'_String) UInt8 : 4
               FUNCTION in(source_type : 2, __set_String_8953384307879712516_13388813323589638527 :: 6) -> in(__table1.source_type, __set_String_8953384307879712516_13388813323589638527) UInt8 : 5
               FUNCTION notEquals(ad_id : 3, \'global\'_String :: 8) -> notEquals(__table1.ad_id, \'global\'_String) UInt8 : 6
               FUNCTION less(toUInt64OrZero(__table1.account_id) :: 9, 900000_UInt32 :: 7) -> less(toUInt64OrZero(__table1.account_id), 900000_UInt32) UInt8 : 8
               FUNCTION and(greaterOrEquals(__table1.day, \'2025-07-01\'_String) :: 10, lessOrEquals(__table1.day, \'2025-07-31\'_String) :: 4, in(__table1.source_type, __set_String_8953384307879712516_13388813323589638527) :: 5, notEquals(__table1.ad_id, \'global\'_String) :: 6, less(toUInt64OrZero(__table1.account_id), 900000_UInt32) :: 8) -> and(greaterOrEquals(__table1.day, \'2025-07-01\'_String), lessOrEquals(__table1.day, \'2025-07-31\'_Stri... UInt8 : 7
      Positions: 0 1 2 3 7

49 rows in set. Elapsed: 0.004 sec. 
```

antaljanosbenjamin committed:
Do not remove inputs from ActionsDAG after `mergeInplace`

Some part of the code might depend on the fact that inputs are not removed from the DAG. However these `INPUT` type actions are not part of the inputs list of the resulting DAG, so removing unused actions should remove them.

antaljanosbenjamin committed:
Merge branch 'master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix filter push-down when the pre join actions remove some of the splitted filter's inputs from their output

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

`JoinStepLogical::removeUnusedColumn` and the tests are not fixed purposefully, wanted to keep the changes separate

antaljanosbenjamin committed:
Reimplement `JoinStepLogical::removeUnusedColumns`

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix rename

antaljanosbenjamin committed:
Fix test after merge

antaljanosbenjamin committed:
Fix fast test compilation

antaljanosbenjamin committed:
Handle removing of all columns from JoinStepLogical properly

antaljanosbenjamin committed:
Style fixes

antaljanosbenjamin committed:
Handle removal of filter column in FilterStep

antaljanosbenjamin committed:
Handle dummy columns from other join steps properly in JoinStepLogical

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fixes after merge

antaljanosbenjamin committed:
Keep the source relations for actions in JoinStepLogical

antaljanosbenjamin committed:
Test fixes

antaljanosbenjamin committed:
Fix input-output header mismatch after pushing down disjunctions for JOINs

antaljanosbenjamin committed:
Keep the input order for JOINs when pushing down disjunctions for JOINs

antaljanosbenjamin committed:
Add extra check for partial join filter pushdown

antaljanosbenjamin committed:
Keep headers consistent during JOIN to IN conversion

antaljanosbenjamin committed:
Declare used error code

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix JOIN source mapping even when outputs are not removed

antaljanosbenjamin committed:
Fix test with old analyzer

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix expected output

devcrafter committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

devcrafter commented on tests/queries/0_stateless/01062_pm_all_join_with_block_continuation.sql:
```diff
@@ -21,11 +21,12 @@ SELECT count(1), uniqExact(n) FROM (
 SET query_plan_join_swap_table = 'false';
 
+-- Because of the optimizations in the analyzer the following queries started to run without issues. To keep the essence of the test, we test both cases.
+SET enable_analyzer = 0;
```
afaiu, we have separate runs with and without analyzer in CI. So, there should be no need to change the test

devcrafter commented on src/Interpreters/JoinExpressionActions.h:
```diff
@@ -153,6 +153,7 @@ class JoinExpressionActions
 
     std::shared_ptr<ActionsDAG> getActionsDAG() const;
 
+    void resetNodeSources(NodeToSourceMapping expression_sources);
```
What are node sources? Hardly could find any comment related to it

devcrafter commented on src/Processors/QueryPlan/Optimizations/removeUnusedColumns.cpp:
```diff
@@ -0,0 +1,196 @@
+        if (!child_step->canRemoveUnusedColumns())
+            continue;
+
+        const auto can_remove_inputs = canAllChildrenCanRemoveOutputs(*child_node);
```
It looks we better apply the optimization while walking bottom up though the plan, - this way we can avoid checking children.

devcrafter commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -509,6 +514,22 @@ ActionsDAG::NodeRawConstPtrs ActionsDAG::findInOutputs(const Names & names) cons
+    for (const auto * output : outputs)
+    {
+        if (auto it = possible_output_names.find(output->result_name); it != possible_output_names.end())
+    {
```
formatting

devcrafter reviewed (commented):
Current comments

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Fix SettingsChangesHistory

antaljanosbenjamin committed:
Make analyzer explicit using only query level settings

antaljanosbenjamin committed:
Fix expected outputs

antaljanosbenjamin committed:
Add and improve some comments

antaljanosbenjamin committed:
Fix expected output for test

antaljanosbenjamin commented on tests/queries/0_stateless/03595_convert_any_join_to_semi_or_anti.reference:
```diff
@@ -8,22 +8,21 @@ Positions: 0
           Expression (Left Pre Join Actions)
           Actions: INPUT :: 0 -> __table1.uid Int16 : 0
-          Positions: 0
+                   COLUMN Const(UInt8) -> __join_result_dummy UInt8 : 1
```
Okay, so basically we are creating `__join_result_dummy` in pre join actions, thus we need this extra column here. Furthermore, the join only output the join result dummy, so it should be fine and shouldn't cause any issues.

antaljanosbenjamin committed:
Fix setting changes history

I messed up the previous one, not sure how.

antaljanosbenjamin committed:
Fix merge tree setting changes history

antaljanosbenjamin committed:
Small review fixes

antaljanosbenjamin committed:
Convert `UnusedColumnRemovalResult` to enum for less bugphrone usage

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

devcrafter commented on src/Processors/QueryPlan/Optimizations/removeUnusedColumns.cpp:
```diff
@@ -0,0 +1,196 @@
+    return RemoveChildrenOutputResult::NotUpdated;
+}
+
+size_t tryRemoveUnusedColumns(QueryPlan::Node * node, QueryPlan::Nodes & nodes, const Optimization::ExtraSettings &)
```
Still would be nice to have details related to the optimization

antaljanosbenjamin replied:
I added a [comment](https://github.com/ClickHouse/ClickHouse/blob/a530f342028cb5b6cb88830db013b1e6ee34a95a/src/Processors/QueryPlan/Optimizations/Optimizations.h#L111-L123) in the `Optimizations.h` header. If you think we need more details, could you please point it out what should I include in the docs?

devcrafter commented on src/Processors/QueryPlan/Optimizations/removeUnusedColumns.cpp:
```diff
@@ -0,0 +1,196 @@
+        const auto updated_anything
+            = child_step->removeUnusedColumns(getNameMultiSetFromNames(parent_inputs[child_id]->getNames()), false).updated_anything;
+
+        // TODO(antaljanosbenjamin): compare the header with name and types
```
remove or add some assert?

antaljanosbenjamin commented on src/Processors/QueryPlan/IQueryPlanStep.h:
```diff
@@ -118,6 +118,28 @@ class IQueryPlanStep
+        NothingChanged,
+        UpdatedButKeptInputs,
+        RemovedInputs
+    };
```
I had the same feeling, but wasn't sure about naming. I love your idea, will rename.

devcrafter reviewed (approved):
LGTM

antaljanosbenjamin committed:
Use shorter name and values for result enum

antaljanosbenjamin committed:
Add concrete example to explaining comment

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Merge branch 'master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Merge branch 'master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into optimize-remove-unused-columns-in-plan

antaljanosbenjamin committed:
Revert redundant fix
